### ISSUE TITLE
Error when registering expansions

### ISSUE THREAD

**Issue by NicBOMB** [PlaceholderAPI/PlaceholderAPI#515]
### Issue
Expansions are not successfully registered after downloading using `papi ecloud download <expansion>`.

### Expected behaviour
PAPI should download, prompt to reload, and register the expansion after reloading.

### Actual behaviour
PAPI downloads the expansion, prompts to reload, but throws null pointer exceptions when reloading.
https://paste.helpch.at/redubilofa.bash

### How to reproduce

1. Download expansion using `papi ecloud download server`
2. `Successfully downloaded expansion Server [2.4.0] to file: Expansion-server.jar
Make sure to type /papi reload to enable your new expansion!`
3. Reload using `papi reload`

### `/papi dump` output
When I've been downloading expansions all names have been "Expansion-[lowercase plugin name]". Reloading PAPI does not register the newly downloaded expansions automatically. After looking at other issues and dumps, I realized that my expansion for server was called "Expansion-server.jar" in my dump, but "Expansion-Server.jar" in another user's dump. I redownloaded server from the ecloud and console still reported
`Successfully downloaded expansion Server [2.4.0] to file: Expansion-server.jar
Make sure to type /papi reload to enable your new expansion!`
First Dump https://paste.helpch.at/obapozepuf
After renaming the downloaded file to "Expansion-Server.jar" and then reloading PAPI, the expansion was successfully registered and recorded in a subsequent dump. https://paste.helpch.at/wesawaseto

Is there a list of the names or syntax PAPI looks for? `papi ecloud list all`? Some don't register using their ecloud name either. Example: "mcMMO" as seen in `papi ecloud list all` isn't registered when named "Expansion-mcmmo.jar" or "Expansion-mcMMO.jar" but is registered when named "Expansion-McMMO.jar".
I cannot get ListPlayers and SpeedPerSec to register. All of the others I have renamed and were successfully registered.
https://paste.helpch.at/olopujunoy
If I remove ListPlayers and SpeedPerSec the above null pointer exception also disappears.
https://paste.helpch.at/ruhefidomo
I'd really like to be able to just download, reload, and move on to implementation. The plugin should get the placeholders for me without needing me to double check the files are named correctly.

**Comment by Andre601** [comment]
Hey there.

Please use the paste-service (https://paste.helpch.at) next time for sharing stacktraces. I gone ahead and quickly did this for you.

About your issue: Usually the name of the expansion-jar doesn't matter as the plugin will just go through the expansions-folder and try to load each jar in it.
The name/identifier that was set within the jar's code and the fact that the jar's main class should extend the PlaceholderExpansion is more important.

Did you try out the latest dev-build and see if this issue was already fixed?

**Comment by NicBOMB** [comment]
> Hey there.
> 
> Please use the paste-service (https://paste.helpch.at) next time for sharing stacktraces. I gone ahead and quickly did this for you.
> 
> About your issue: Usually the name of the expansion-jar doesn't matter as the plugin will just go through the expansions-folder and try to load each jar in it.
> The name/identifier that was set within the jar's code and the fact that the jar's main class should extend the PlaceholderExpansion is more important.
> 
> Did you try out the latest dev-build and see if this issue was already fixed?
Tried the dev build just now.
https://paste.helpch.at/exiqahuhih.md
Same null pointer exception. 
Dump https://paste.helpch.at/lurugujije
ListPlayers and SpeedPerSec are still not registered when downloaded by the latest dev build.

**Comment by Andre601** [comment]
Okay. From your first link can I already see a few minor issues:
- `[17:52:39 WARN]: [PlaceholderAPI] listplayers is attempting to register placeholders via a PlaceholderHook class which is no longer supported!`
The Listplayers expansion is using outdated methods to register itself which isn't supported anymore, so contact the author of it to inform them about the requirement of updating the expansion.
- mcmmo has been registered twice... This is odd and shouldn't happen. Perhaps one is an expansion and the other inbuild placeholders by the plugin?

Considering that the plugin is throwing this error after registering most of the expansions successfully would I assume that perhaps one specific expansion could cause this.

Would it be possible that you could try out a similar server setup, but only with PAPI installed? There you should try to install all the expansions you download one by one and reload in between to see if it breaks something. That's pretty much the only way to test out if the cause is an expansion.
If nothing happens with all the expansions, try to install each of the plugins (which are hooking into PAPI, see the /papi list output and compare each with the expansions you downloaded) and restart the server in between.

Sorry for those annoying and tedious tasks.

**Comment by NicBOMB** [comment]
* `[17:52:39 WARN]: [PlaceholderAPI] listplayers is attempting to register placeholders via a PlaceholderHook class which is no longer supported!`
>       The Listplayers expansion is using outdated methods to register itself which isn't supported anymore, so contact the author of it to inform them about the requirement of updating the expansion.

I will use the world expansion instead. It seems to offer the same placeholders and a few more, so not having ListPlayers is fine. I can't find where I'd make an issue for the author of ListPlayers either, so asking me to won't get that anywhere /s. Either way, I removed listplayers and won't need it since the World expansion registered successfully and provides similar placeholders. 
tldr: #456 duplicate

>     * mcmmo has been registered twice... This is odd and shouldn't happen. Perhaps one is an expansion and the other inbuild placeholders by the plugin?
https://paste.helpch.at/erisidofeh.css
I don't know why it was registering twice. I was running the latest mcmmo (2.1.163), but after searching through the logs it doesn't seem to register any placeholders of its own and only a single `[PlaceholderAPI] Unregistered placeholder expansion: mcmmo` message is ever written despite the duplicate. Searching through the mcmmo discord I found messages saying there is a PAPI expansion and everything indicated it worked for them, but nothing about built-in PAPI placeholders with the mcmmo plugin. Maybe mcmmo was registered twice because one of the other 2 expansions was unable to register successfully? Ultimately I was able to find the dev builds for the mcmmo expansion and after installing "mcmmo-placeholders-plugin-1.0.2-SNAPSHOT.jar" as a plugin instead of as an expansion, it registered successfully and never duplicated. I have yet to test if the dev build placeholders hook into the latest version of mcmmo correctly, but that's a problem for the expansion author if I run into problems, not for here, so we can say mcmmo, its ecloud expansion, and its expansion's dev builds are not the cause and probably work given the positive feedback other users gave.

> Considering that the plugin is throwing this error after registering most of the expansions successfully would I assume that perhaps one specific expansion could cause this.
> 
> Would it be possible that you could try out a similar server setup, but only with PAPI installed? There you should try to install all the expansions you download one by one and reload in between to see if it breaks something. That's pretty much the only way to test out if the cause is an expansion.
> If nothing happens with all the expansions, try to install each of the plugins (which are hooking into PAPI, see the /papi list output and compare each with the expansions you downloaded) and restart the server in between.
> 
> Sorry for those annoying and tedious tasks.

Doing this last part, everything was fine till I installed speedpersec AND listplayers. They both seem to cause null pointer exceptions individually. It seems that plugins not registered before those two are not registered after the error. McMMO happens to be the unregistered placeholder in this example: https://paste.helpch.at/ijuferozux.md https://paste.helpch.at/hugicovare
Those two are definitely causing a problem for PAPI, but PAPI also isn't handling those errors well. The only solution seems to be to not use those expansions till they are fixed. I'll create an issue for speedpersec.

**Comment by Andre601** [comment]
Really odd...
The error is pointing to line 343 in the LocalExpansionManager which is this one here if I use the right version of it:
https://github.com/PlaceholderAPI/PlaceholderAPI/blob/51844cd291bcac0b2a8c6b797e51eaefbac8e916/src/main/java/me/clip/placeholderapi/expansion/manager/LocalExpansionManager.java#L343

So it could be an issue with the Logging at that point.
Could you try it out with the latest dev build and give me the error from that one, so that I can check the actual code for it? I'll for sure update this odd way of logging stuff.

**Comment by NicBOMB** [comment]
> Really odd...
> The error is pointing to line 343 in the LocalExpansionManager which is this one here if I use the right version of it:
> 
> https://github.com/PlaceholderAPI/PlaceholderAPI/blob/51844cd291bcac0b2a8c6b797e51eaefbac8e916/src/main/java/me/clip/placeholderapi/expansion/manager/LocalExpansionManager.java#L343
> 
> So it could be an issue with the Logging at that point.
> Could you try it out with the latest dev build and give me the error from that one, so that I can check the actual code for it? I'll for sure update this odd way of logging stuff.

I just tried with the latest dev build. Dumps are included in log below.
https://paste.helpch.at/odogiyifab.md

**Comment by Andre601** [comment]
As I expected, it is related to the logging:
https://github.com/PlaceholderAPI/PlaceholderAPI/blob/master/src/main/java/me/clip/placeholderapi/expansion/manager/LocalExpansionManager.java#L370

Problem is most likely that `clazz` is null here, so I'll try to find something to fix this...

**Comment by Andre601** [comment]
The issue is found.
Apparently is the speedpersec expansion's jar malformed. It doesn't contain any class file whatsoever which makes PlaceholderAPI fail to load it.

I'll still fix the logging issue as such a case shouldn't make the plugin fail to log info properly, but the main issue is caused by the expansion itself, so better let the dev know about this.

**Comment by PiggyPiglet** [comment]
reopening cuz issue was only partially fixed
